### ISSUE TITLE
OpenModelPhase cannot go past 1000 rps

### ISSUE THREAD

**Issue by franz1981** [Hyperfoil/Hyperfoil#342]
Due to the millisecond delta precision, users cannot exceed 1000 rps for constantRate, rumpUp-kind of workload

**Comment by franz1981** [comment]
See https://github.com/Hyperfoil/Hyperfoil/blob/5dbd941d579c47c1f5a03021468f266934ba9e80/core/src/main/java/io/hyperfoil/core/impl/PhaseInstanceImpl.java#L350

**Comment by franz1981** [comment]
It would be nice to make uses of a GC-free microseconds time from epoch timer, if possible, too, without using JNI/JNA.
The simpler approach is to adapt https://github.com/apache/activemq-artemis/blob/b9bb494b580ce59d41e63d4cd93ffdb07dc5dfaa/artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/perf/MicrosClock.java#L23

Which relies on https://github.com/openjdk/jdk/blob/4785461f61d8f5c7444d2e6fd90f1e083dbc6fe4/src/hotspot/share/prims/jvm.cpp#L258 but I am not sure it is reliable across machines, if compared to other, more atomic ways.

**Comment by franz1981** [comment]
this still means @wburns this is no longer a blocker for you, wdyt?

**Comment by willr3** [comment]
I think I follow along the with the thought process. Are we seeing that it loops 1000 times at the beginning of each millisecond tick and instead we need it to spread the 1000 new requests over the full millisecond?

**Comment by franz1981** [comment]
Exactly, beacuse right now, given that we can schedule the next fire time only at millisecond granularity, we just schedule the batch required to cover for the current missing ones (assuming we're keeping up with the avg expected rate)

i.e. 100K rps -> 100 rpms (rate per millisecond)

**Comment by wburns** [comment]
This sounds like it could lead to spikey latencies if the user/second is a good chunk more than 1K/s. For example if your goal was 10K/s that means you will have 10 requests immediately at the beginning of a millisecond? In some of my tests my average latency is < 200 us so that would be make a big impact, where my client may be slammed for 200-400 us and then doing nothing for the remainder of that millisecond?

Also I assume this doesn't affect the number of agents. That is the 1K/s is only per agent, so technically you can do up to 1000 * N operations/s per agent and not see this behavior?

**Comment by franz1981** [comment]
The answer here is more complex than it seems: it depends by the available cores to Hyperfoil and how the sessions are distributed among event loops (which should be a fair distribution).

TLDR OpenModelPhase just schedule new sessions, and:
- If there is a single event loop they are going to be enqueued and executed after the tick (and acquisition from the session pool) complete
- if there are more, they will get (at best) M sessions % M event loops long queues of sessions to be executed per event loop, but most importantly, will fire asap, still getting some delay due to session pool acquisition order 


None of these scheduling happen instantaneously, but we need to measure the infrastructure cost under such level of load, IMO, to make sure of it.

IMO we could achieve sub millis scheduling from Netty with some better interleaving between sessions run and scheduling's decisions, but the only way to make sure is by adding telemetry to the OpenModelPhase itself, possibly in some micro benchmarks
